### PR TITLE
Read asynchronously from test host process.

### PR DESCRIPTION
**Pull request by mayankbansal018**

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 1]
```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
+                                {
+                                    testHostProcessStdError.Remove(0, required);
+                                }
+                            }
+
+                            testHostProcessStdError.Append(data);
                         }
-
-                        this.RequestSender.OnClientProcessExit(testHostProcessStdError);
+                        if (process.HasExited && process.ExitCode != 0)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: `process.HasExited` can throw InvalidOperationException if no process associated with object,  `process.ExitCode` can throw IOE if the process has not exited. is there a guaranty process.HasExited doesn't throw in callback?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 2]
```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
 
                 if (testHostStartInfo != null)
                 {
-                    // Monitor testhost exit.
-                    testHostStartInfo.ExitCallback = (process) =>
+                    // Monitor testhost error callbacks.
+                    testHostStartInfo.ErrorReceivedCallback = (process, data) =>
                     {
-                        if (process.ExitCode != 0)
+                        if(data != null)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Nit: move this to a method.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 1 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Tested process.HasExited out, the process object is valid, even after the actual process has ended.
Process.ExitCode is only called if the process.HasExited is successful.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 2 above)
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: nit: `if (` instead `if(`.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 3]
```diff
@@ -146,5 +170,14 @@ private string GetTimestampedLogFile(string logFile)
                 string.Format("host.{0}_{1}{2}", DateTime.Now.ToString("yy-MM-dd_HH-mm-ss_fffff"),
                     Thread.CurrentThread.ManagedThreadId, Path.GetExtension(logFile)));
         }
+
+        /// <summary>
+        /// Returns the current error data in stream
+        /// Written purely for UT as of now.
+        /// </summary>
+        public virtual string GetStandardError()
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Make it protected. Use a `Testable` in unit test.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 4]
```diff
@@ -39,12 +44,14 @@ public abstract class ProxyOperationManager
         /// <param name="requestSender">Request Sender instance.</param>
         /// <param name="testHostManager">Test host manager instance.</param>
         /// <param name="clientConnectionTimeout">Client Connection Timeout.</param>
-        protected ProxyOperationManager(ITestRequestSender requestSender, ITestHostManager testHostManager, int clientConnectionTimeout)
+        protected ProxyOperationManager(ITestRequestSender requestSender, ITestHostManager testHostManager, int clientConnectionTimeout, int errorLength = 1000)
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Explicit arguments are better. Please call with `1000` from the public ctor.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 5]
```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
-                        if (process.ExitCode != 0)
+                        if(data != null)
                         {
-                            testHostProcessStdError = process.StandardError.ReadToEnd();
-                            EqtTrace.Error("Test host exited with error: {0}", testHostProcessStdError);
+                            //if incoming data stream is huge empty entire testError stream, & limit data stream to MaxCapacity
+                            if (data.Length > testHostProcessStdError.MaxCapacity)
+                            {
+                                testHostProcessStdError.Remove(0, testHostProcessStdError.Length);
+                                data = data.Substring(data.Length - testHostProcessStdError.MaxCapacity);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: We're keeping the `first` few characters, should it be `last` few characters?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 6]
```diff
@@ -39,12 +44,14 @@ public abstract class ProxyOperationManager
...
         /// <param name="testHostManager">Test host manager instance.</param>
         /// <param name="clientConnectionTimeout">Client Connection Timeout.</param>
-        protected ProxyOperationManager(ITestRequestSender requestSender, ITestHostManager testHostManager, int clientConnectionTimeout)
+        protected ProxyOperationManager(ITestRequestSender requestSender, ITestHostManager testHostManager, int clientConnectionTimeout, int errorLength = 1000)
         {
             this.RequestSender = requestSender;
             this.connectionTimeout = clientConnectionTimeout;
             this.testHostManager = testHostManager;
             this.processHelper = new ProcessHelper();
+            this.errorLength = errorLength;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Instance variable is not used any where?

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 7]
```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
-                            testHostProcessStdError = process.StandardError.ReadToEnd();
-                            EqtTrace.Error("Test host exited with error: {0}", testHostProcessStdError);
+                            //if incoming data stream is huge empty entire testError stream, & limit data stream to MaxCapacity
+                            if (data.Length > testHostProcessStdError.MaxCapacity)
+                            {
+                                testHostProcessStdError.Remove(0, testHostProcessStdError.Length);
+                                data = data.Substring(data.Length - testHostProcessStdError.MaxCapacity);
+                            }
+
+                            //remove only what is required, from begining of error stream
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: nit: spelling

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 8]
```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
...
                     {
-                        if (process.ExitCode != 0)
+                        if(data != null)
                         {
-                            testHostProcessStdError = process.StandardError.ReadToEnd();
-                            EqtTrace.Error("Test host exited with error: {0}", testHostProcessStdError);
+                            //if incoming data stream is huge empty entire testError stream, & limit data stream to MaxCapacity
+                            if (data.Length > testHostProcessStdError.MaxCapacity)
+                            {
+                                testHostProcessStdError.Remove(0, testHostProcessStdError.Length);
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Which of these is most efficient?
1. StringBuilder.Clear followed by StringBuilder.Append
2. StringBuilder.Remove(x, y) followed by StringBuilder.Append
3. StringBuilder[i] = data[i]

File: src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs [hunk 9]
```diff
@@ -39,13 +39,15 @@ public Process LaunchProcess(string processPath, string arguments, string workin
...
+                if (errorCallback != null)
                 {
-                    process.Exited += (sender, args) => exitCallback(sender as Process);
+                    process.ErrorDataReceived += (sender, args) => errorCallback(sender as Process, args.Data);
                 }
 
                 EqtTrace.Verbose("ProcessHelper: Starting process '{0}' with command line '{1}'", processPath, arguments);
                 process.Start();
+
+                process.BeginErrorReadLine();
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Helpers/ProcessHelper.cs]: Do we require this even if `errorCallback` is not set?

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Helpers/ProcessHelperTests.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+            this.testOperationManager.SetupChannel(Enumerable.Empty<string>());
+
+            Assert.AreEqual(this.testOperationManager.GetStandardError(), errorData);
+        }
+
+        [TestMethod]
+        public void ErrorMessageShouldBeTruncatedToMatchErrorLength()
+        {
+            string errorData = "Long Custom Error Strings";
```
Comment (codito) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Helpers/ProcessHelperTests.cs]: Suggest to organize unit tests as follows for readability (comments are not required, newlines are only between blocks):
```
// Block of arrange
int x = 10;
double y = 9.0

// Block of act
CallFoo(y);

// Block of assert
Assert(12.0, y);
Assert(29, x);
```

**Review by codito — APPROVED** [review]
Looks good with few suggested changes.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs (hunk 5 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: we are keeping last few only substring(index) which will give from starting index to end

**Comment by mayankbansal018** [comment]
Bug: https://github.com/Microsoft/vstest/issues/521

**Pull request merged**